### PR TITLE
Add tests for Java 8 TYPE_USE annotations on arrays (#614).

### PR DESCRIPTION
**Pull request by rluble**

### CODE REVIEW

**Comment by rluble** [comment]
Yes I saw the way it applies to varargs. I'll add the tests. I guess they chose the ordering for the brackets on array types that way to match the ordering on array access and array instantiation where the first bracket is the outermost dimension, and that carried over to varargs which is a bit unnatural.

File: src/test/java/com/squareup/javapoet/AnnotatedTypeNameTest.java [hunk 1]
```diff
@@ -137,4 +140,79 @@ private void annotatedEquivalence(TypeName type) {
     String actual = type.toString();
     assertEquals(expected, actual);
   }
+
+  // https://github.com/square/javapoet/issues/614
+  @Ignore @Test public void annotatedArrayType() {
+    String expected =
+        "java.lang.Object @" + TypeUseAnnotation.class.getCanonicalName() + " []";
```
Comment (tbroyer) [inline src/test/java/com/squareup/javapoet/AnnotatedTypeNameTest.java]: Should this introduce `TUA` and `TYPE_USE_ANNOTATION` constants like the `NN` and `NEVER_NULL` above? That would probably make the tests a bit easier to read. WDYT?

Comment (rluble) [reply src/test/java/com/squareup/javapoet/AnnotatedTypeNameTest.java]: Good idea.

**Comment by rluble** [comment]
I was looking into fixing the whole type use annotations and it requires a bit of refactoring the design. Currently the way annotations on types are handled is already quite hacky IMHO. Look at the difference between `TypeName.emit` and `TypeName.toString` which blur the line of where the responsibility for emitting annotations lie.

The only concern is that currently type annotations are treated as if they were annotations on the enclosing construct (`ParameterSpec`, etc etc) for most part. Fixing the issue might break questionable uses, i.e. someone using a `ElementType.METHOD` annotation on the return type of a method.

**Comment by rluble** [comment]
I'll have to abandon this pull request and submit a new. It seems that I should be using my @google.com email instead of my personal email for the commits.

**Comment by rluble** [comment]
Let's continue the review at #618

**Pull request closed without merging**